### ISSUE TITLE
tensorflow.lite.python.convert.ConverterError: <unknown>:0: error: loc("Func/StatefulPartitionedCall/input/_0"): requires all operands and results to have compatible element types

### ISSUE THREAD

**Issue by clgilbe** [tensorflow/tensorflow#42079]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows
- TensorFlow installed from (source or binary): Anaconda
- TensorFlow version (or github SHA if from source):  tf_nightly_gpu-2.4.0.dev20200805-cp37-cp37m-win_amd64.whl

**Command used to run the converter or code if you’re using the Python API**
If possible, please share a link to Colab/Jupyter/any notebook.

```
model = tf.saved_model.load(saved_model_dir)
concrete_func = model.signatures[tf.saved_model.DEFAULT_SERVING_SIGNATURE_DEF_KEY]
concrete_func.inputs[0].set_shape([1, 300, 300, 3])
converter = tf.lite.TFLiteConverter.from_concrete_functions([concrete_func])


# These lines are necessary for the issue fix https://github.com/tensorflow/tensorflow/issues/41877
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.experimental_new_converter = True
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
tflite_model = converter.convert()
```

**The output from the converter invocation**

```
(tf_gpu_tf_nightly) d:\TensorFlow\tensorflowComponentId\models\research\object_detection>python convert_saved_model_to_f
lat_buffer.py
2020-08-05 16:31:38.096415: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cudart64_101.dll
2020-08-05 16:31:46.030792: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library nvcuda.dll
2020-08-05 16:31:46.153774: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: Quadro P4000 computeCapability: 6.1
coreClock: 1.48GHz coreCount: 14 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 226.62GiB/s
2020-08-05 16:31:46.176284: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cudart64_101.dll
2020-08-05 16:31:46.371687: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cublas64_10.dll
2020-08-05 16:31:46.632753: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cufft64_10.dll
2020-08-05 16:31:46.734036: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library curand64_10.dll
2020-08-05 16:31:47.053455: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cusolver64_10.dll
2020-08-05 16:31:47.157739: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cusparse64_10.dll
2020-08-05 16:31:47.917165: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cudnn64_7.dll
2020-08-05 16:31:47.930832: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0
2020-08-05 16:31:47.966000: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized wit
h oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:
 AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-08-05 16:31:48.030523: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x214141cae40 initialized for
platform Host (this does not guarantee that XLA will be used). Devices:
2020-08-05 16:31:48.088275: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default
 Version
2020-08-05 16:31:48.136648: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: Quadro P4000 computeCapability: 6.1
coreClock: 1.48GHz coreCount: 14 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 226.62GiB/s
2020-08-05 16:31:48.156650: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cudart64_101.dll
2020-08-05 16:31:48.166699: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cublas64_10.dll
2020-08-05 16:31:48.176341: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cufft64_10.dll
2020-08-05 16:31:48.187558: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library curand64_10.dll
2020-08-05 16:31:48.198596: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cusolver64_10.dll
2020-08-05 16:31:48.207628: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cusparse64_10.dll
2020-08-05 16:31:48.219176: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cudnn64_7.dll
2020-08-05 16:31:48.230180: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0
2020-08-05 16:31:48.869564: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1257] Device interconnect StreamExecutor
with strength 1 edge matrix:
2020-08-05 16:31:48.878309: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1263]      0
2020-08-05 16:31:48.883346: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1276] 0:   N
2020-08-05 16:31:48.888900: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/job:loc
alhost/replica:0/task:0/device:GPU:0 with 6307 MB memory) -> physical GPU (device: 0, name: Quadro P4000, pci bus id: 00
00:01:00.0, compute capability: 6.1)
2020-08-05 16:31:48.914058: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x21455b650a0 initialized for
platform CUDA (this does not guarantee that XLA will be used). Devices:
2020-08-05 16:31:48.924375: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Quadro P4000,
 Compute Capability 6.1
2020-08-05 16:32:21.722971: I tensorflow/core/grappler/devices.cc:69] Number of eligible GPUs (core count >= 8, compute
capability >= 0.0): 1
2020-08-05 16:32:21.732162: I tensorflow/core/grappler/clusters/single_machine.cc:356] Starting new session
2020-08-05 16:32:21.738833: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: Quadro P4000 computeCapability: 6.1
coreClock: 1.48GHz coreCount: 14 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 226.62GiB/s
2020-08-05 16:32:21.757141: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cudart64_101.dll
2020-08-05 16:32:21.764885: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cublas64_10.dll
2020-08-05 16:32:21.775482: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cufft64_10.dll
2020-08-05 16:32:21.783435: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library curand64_10.dll
2020-08-05 16:32:21.791890: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cusolver64_10.dll
2020-08-05 16:32:21.799672: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cusparse64_10.dll
2020-08-05 16:32:21.813984: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic
library cudnn64_7.dll
2020-08-05 16:32:21.822129: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0
2020-08-05 16:32:21.827773: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1257] Device interconnect StreamExecutorwith strength 1 edge matrix:
2020-08-05 16:32:21.839959: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1263]      0
2020-08-05 16:32:21.844847: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1276] 0:   N
2020-08-05 16:32:21.850558: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6307 MB memory) -> physical GPU (device: 0, name: Quadro P4000, pci bus id: 00
00:01:00.0, compute capability: 6.1)
2020-08-05 16:32:23.381950: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:832] Optimization results for grappler item: graph_to_optimize
2020-08-05 16:32:23.392215: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:834]   function_optimizer: Graph size after: 9459 nodes (8971), 11780 edges (11285), time = 283.837ms.
2020-08-05 16:32:23.408641: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:834]   function_optimizer: function_optimizer did nothing. time = 6.568ms.
2020-08-05 16:32:44.552016: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:315] Ignored output_format.
2020-08-05 16:32:44.561287: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:318] Ignored drop_control_dependency.
loc("Func/StatefulPartitionedCall/input/_0"): error: requires all operands and results to have compatible element types
Traceback (most recent call last):
  File "C:\Users\clgilbe\Anaconda3\envs\tf_gpu_tf_nightly\lib\site-packages\tensorflow\lite\python\convert.py", line 199, in toco_convert_protos
    enable_mlir_converter)
  File "C:\Users\clgilbe\Anaconda3\envs\tf_gpu_tf_nightly\lib\site-packages\tensorflow\lite\python\wrap_toco.py", line 38, in wrapped_toco_convert
    enable_mlir_converter)
Exception: <unknown>:0: error: loc("Func/StatefulPartitionedCall/input/_0"): requires all operands and results to have compatible element types
<unknown>:0: note: loc("Func/StatefulPartitionedCall/input/_0"): see current operation: %1 = "tf.Identity"(%arg0) {device = ""} : (tensor<1x300x300x3x!tf.quint8>) -> tensor<1x300x300x3xui8>


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "convert_saved_model_to_flat_buffer.py", line 21, in <module>
    tflite_model = converter.convert()
  File "C:\Users\clgilbe\Anaconda3\envs\tf_gpu_tf_nightly\lib\site-packages\tensorflow\lite\python\lite.py", line 1118,in convert
    return super(TFLiteConverterV2, self).convert()
  File "C:\Users\clgilbe\Anaconda3\envs\tf_gpu_tf_nightly\lib\site-packages\tensorflow\lite\python\lite.py", line 942, in convert
    self).convert(graph_def, input_tensors, output_tensors)
  File "C:\Users\clgilbe\Anaconda3\envs\tf_gpu_tf_nightly\lib\site-packages\tensorflow\lite\python\lite.py", line 669, in convert
    **converter_kwargs)
  File "C:\Users\clgilbe\Anaconda3\envs\tf_gpu_tf_nightly\lib\site-packages\tensorflow\lite\python\convert.py", line 574, in toco_convert_impl
    enable_mlir_converter=enable_mlir_converter)
  File "C:\Users\clgilbe\Anaconda3\envs\tf_gpu_tf_nightly\lib\site-packages\tensorflow\lite\python\convert.py", line 202, in toco_convert_protos
    raise ConverterError(str(e))
tensorflow.lite.python.convert.ConverterError: <unknown>:0: error: loc("Func/StatefulPartitionedCall/input/_0"): requires all operands and results to have compatible element types
<unknown>:0: note: loc("Func/StatefulPartitionedCall/input/_0"): see current operation: %1 = "tf.Identity"(%arg0) {device = ""} : (tensor<1x300x300x3x!tf.quint8>) -> tensor<1x300x300x3xui8>
```

**Also, please include a link to the saved model or GraphDef**

```
Pre-trained model ssd_resnet50_v1_fpn_640x640_coco17_tpu-8 http://download.tensorflow.org/models/object_detection/tf2/20200711/ssd_resnet50_v1_fpn_640x640_coco17_tpu-8.tar.gz
```

**Failure details**
Failure to convert using instructions for converting a saved model using concrete_functions.  https://www.tensorflow.org/lite/convert/python_api#converting_a_savedmodel_  Similar error to #41877

**RNN conversion support**
If converting TF RNN to TFLite fused RNN ops, please prefix [RNN] in the title. - NOPE

**Any other info / logs**

See above

**Comment by Saduf2019** [comment]
@clgilbe 
I ran your code with the changes as per #41877 and do not face any error, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/aa58a668ec9ddff9339fa5a12b7f754a/untitled330.ipynb#scrollTo=-COf3oVok4lC)

**Comment by hahmad2008** [comment]
@Saduf2019 
Hi I tried same code you shared, but I got this exception:

----
**!pip install tf-nightly**
```
import tensorflow as tf
saved_model_dir='ssd_mobilenet_v2_320x320_coco17_tpu-8/saved_model/'

model = tf.saved_model.load(saved_model_dir)
concrete_func = model.signatures[tf.saved_model.DEFAULT_SERVING_SIGNATURE_DEF_KEY]
concrete_func.inputs[0].set_shape([1, 300, 300, 3])
converter = tf.lite.TFLiteConverter.from_concrete_functions([concrete_func])


# These lines are necessary for the issue fix https://github.com/tensorflow/tensorflow/issues/41877
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.experimental_new_converter = True
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]

tflite_model = converter.convert()
open('tflite_model_coco17o.tflite','wb').write(tflite_model)
```

---

> ConverterError: <unknown>:0: error: loc("Func/StatefulPartitionedCall/input/_0"): requires all operands and results to have compatible element types
> <unknown>:0: note: loc("Func/StatefulPartitionedCall/input/_0"): see current operation: %1 = "tf.Identity"(%arg0) {device = ""} : (tensor<1x300x300x3x!tf.quint8>) -> tensor<1x300x300x3xui8>
> 

```
Exception                                 Traceback (most recent call last)
/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/convert.py in toco_convert_protos(model_flags_str, toco_flags_str, input_data_str, debug_info_str, enable_mlir_converter)
    198                                                  debug_info_str,
--> 199                                                  enable_mlir_converter)
    200       return model_str

6 frames
/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/wrap_toco.py in wrapped_toco_convert(model_flags_str, toco_flags_str, input_data_str, debug_info_str, enable_mlir_converter)
     37       debug_info_str,
---> 38       enable_mlir_converter)
     39 

Exception: <unknown>:0: error: loc("Func/StatefulPartitionedCall/input/_0"): requires all operands and results to have compatible element types
<unknown>:0: note: loc("Func/StatefulPartitionedCall/input/_0"): see current operation: %1 = "tf.Identity"(%arg0) {device = ""} : (tensor<1x300x300x3x!tf.quint8>) -> tensor<1x300x300x3xui8>


During handling of the above exception, another exception occurred:

ConverterError                            Traceback (most recent call last)
<ipython-input-9-2ee2db71f5c7> in <module>()
----> 1 tflite_model = converter.convert()
      2 open('tflite_model_coco17o.tflite','wb').write(tflite_model)

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/lite.py in convert(self)
   1116         Invalid quantization parameters.
   1117     """
-> 1118     return super(TFLiteConverterV2, self).convert()
   1119 
   1120 

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/lite.py in convert(self)
    940 
    941     return super(TFLiteFrozenGraphConverterV2,
--> 942                  self).convert(graph_def, input_tensors, output_tensors)
    943 
    944 

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/lite.py in convert(self, graph_def, input_tensors, output_tensors)
    667         input_tensors=input_tensors,
    668         output_tensors=output_tensors,
--> 669         **converter_kwargs)
    670 
    671     calibrate_and_quantize, flags = quant_mode.quantizer_flags(

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/convert.py in toco_convert_impl(input_data, input_tensors, output_tensors, enable_mlir_converter, *args, **kwargs)
    572       input_data.SerializeToString(),
    573       debug_info_str=debug_info_str,
--> 574       enable_mlir_converter=enable_mlir_converter)
    575   return data
    576 

/usr/local/lib/python3.6/dist-packages/tensorflow/lite/python/convert.py in toco_convert_protos(model_flags_str, toco_flags_str, input_data_str, debug_info_str, enable_mlir_converter)
    200       return model_str
    201     except Exception as e:
--> 202       raise ConverterError(str(e))
    203 
    204   if distutils.spawn.find_executable(_toco_from_proto_bin) is None:

ConverterError: <unknown>:0: error: loc("Func/StatefulPartitionedCall/input/_0"): requires all operands and results to have compatible element types
<unknown>:0: note: loc("Func/StatefulPartitionedCall/input/_0"): see current operation: %1 = "tf.Identity"(%arg0) {device = ""} : (tensor<1x300x300x3x!tf.quint8>) -> tensor<1x300x300x3xui8>
```

**Comment by clgilbe** [comment]
@Saduf2019 I agree with @hahmad2008 The error happens on the convert step which I did not see in the gist

**Comment by MeghnaNatraj** [comment]
@clgilbe  can you try the solution posted here? https://github.com/tensorflow/tensorflow/issues/42114#issuecomment-671593386

(save the model and load it again after modifying the input shape?)

**Comment by clgilbe** [comment]
@MeghnaNatraj The suggested solution worked for me to modify the input shape.  Thank you.

**Comment by Saduf2019** [comment]
@clgilbe 
Please move the issue to resolved status if resolved.

**Comment by clgilbe** [comment]
Resolved with solution from #42114 (comment)

**Comment by MeghnaNatraj** [comment]
UPDATE:

With the latest TF Nightly build, conversion works directly without having to update the model

```
# Install TF Nightly (latest TF version which is ongoing development)
# TODO: Restart your runtime after running the following command
!pip install tf-nightly

# Import commands
import tensorflow as tf

# Convert the TF model to TF Lite model
# TODO: Update the path to the model as given below
model_dir = "centernet_resnet50_v1_fpn_512x512_kpts_coco17_tpu-8/saved_model"
converter = tf.lite.TFLiteConverter.from_saved_model(model_dir, signature_keys=['serving_default'])
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
tflite_model = converter.convert()
```

*Note: The model 'tf.ConcatV2' op which is not available in TFLite but it is defined in TF. See instructions: https://www.tensorflow.org/lite/guide/ops_select, which provides details on how you can run inference on this model.*
